lib/deploy/data.ts

```diff
@@ -0,0 +1,296 @@
+        logger.error(e.message);
+        throw e;
+    }
+    if (!data[sdmPackK8s]) {
+        return undefined;
+    }
+    return data[sdmPackK8s];
+}
+
+/**
+ * Given the project, event, [[KubernetesDeployment]] goal, and
+ * handler context generate a default [[KubernetesApplication]] object
+ * that should be suitable for deploying simple, microservices.
+ *
+ * If the SDM configuration contains a `k8s.app` property, i.e.,
+ * `k8Deploy.sdm.configuration.sdm.k8s.app`, it is used as the basis
+ * for the default [[KubernetesApplication]] object.
+ *
+ * This function reads partial Kubernetes specs from the
+ * `.atomist/kubernetes` directory of the project, if any exist.
+ * Specifically it looks for JSON and YAML files with the base names
+ * "deployment", "service", "ingress", "role", "service-account", and
+ * "role-binding".
+ *
+ * It also uses [[defaultImage]] and [[dockerPort]] to further
+ * populated default property values from information in the
+ * repository and event.
+ *
+ * Any values obtained from the repository/project/event/goal override
+ * those in the SDM configuration `k8s` property.
+ *
+ * @param p Project being deployed
+ * @param goalEvent SDM Kubernetes deployment goal event
+ * @param k8Deploy Kubernetes deployment goal configuration
+ * @param context Handler context
+ * @return a valid default KubernetesApplication for this SDM goal deployment event
+ */
+export async function defaultKubernetesApplication(
+    p: GitProject,
+    goalEvent: SdmGoalEvent,
+    k8Deploy: KubernetesDeploy,
+    context: HandlerContext,
+): Promise<KubernetesApplication> {
+
+    const configAppData: Partial<KubernetesApplication> = _.get(k8Deploy, "sdm.configuration.sdm.k8s.app", {});
+
+    const workspaceId = context.workspaceId;
+    const name = goalEvent.repo.name;
+    const ns = configAppData.ns || defaultNamespace;
+    const image = await defaultImage(goalEvent, k8Deploy, context);
+    const port = await dockerPort(p);
+
+    const deploymentSpec: DeepPartial<k8s.V1Deployment> = await loadKubernetesSpec(p, "deployment");
+    const serviceSpec: DeepPartial<k8s.V1Service> = await loadKubernetesSpec(p, "service");
+    const ingressSpec: DeepPartial<k8s.V1beta1Ingress> = await loadKubernetesSpec(p, "ingress");
+    const roleSpec: DeepPartial<k8s.V1Role> = await loadKubernetesSpec(p, "role");
+    const serviceAccountSpec: DeepPartial<k8s.V1ServiceAccount> = await loadKubernetesSpec(p, "service-account");
+    const roleBindingSpec: DeepPartial<k8s.V1RoleBinding> = await loadKubernetesSpec(p, "role-binding");
+
+    const repoAppData: KubernetesApplication = {
+        workspaceId,
+        name,
+        ns,
+        image,
+        port,
+        deploymentSpec,
+        serviceSpec,
+        ingressSpec,
+        roleSpec,
+        serviceAccountSpec,
+        roleBindingSpec,
+    };
+
+    return _.merge(configAppData, repoAppData);
```

Comment:
This mutates the initial `configAppData` from `sdm.configuration.sdm.k8s.app`. Is that what you want?

Reply:
That could be problematic.